File: picker/src/main/java/com/ds/magicpicker/PermissionUtil.java
package com.ds.magicpicker;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;

/**
 * Created by deepak on 14/3/18.
 */

public class PermissionUtil {
    public static final int PERM_READ_EXTERNAL_STORAGE = 121;
    public static final int PERM_WRITE_EXTERNAL_STORAGE = 122;
    public static final int PERM_CAMERA = 123;

    public static boolean checkStoragePermission(final Context context) {
        int currentAPIVersion = Build.VERSION.SDK_INT;
        if(currentAPIVersion>= Build.VERSION_CODES.M) {
            if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.shouldShowRequestPermissionRationale((Activity) context, Manifest.permission.READ_EXTERNAL_STORAGE)) {
                    AlertDialog.Builder alertBuilder = new AlertDialog.Builder(context);
                    alertBuilder.setCancelable(false);
                    alertBuilder.setTitle(context.getString(R.string.permission_required));
                    alertBuilder.setMessage(context.getString(R.string.gallery_permission_required));
                    alertBuilder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            ActivityCompat.requestPermissions((Activity) context, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, PERM_READ_EXTERNAL_STORAGE);
                        }
                    });
                    AlertDialog alert = alertBuilder.create();
                    alert.show();
                } else {
                    ActivityCompat.requestPermissions((Activity) context, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, PERM_READ_EXTERNAL_STORAGE);
                }
                return false;
            } else {
                return checkStorageWritePermission(context);
            }
        } else {
            return true;
        }
    }

    public static boolean checkCameraPermission(final Context context)
    {
        int currentAPIVersion = Build.VERSION.SDK_INT;
        if(currentAPIVersion>= Build.VERSION_CODES.M) {
            if (ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.shouldShowRequestPermissionRationale((Activity) context, Manifest.permission.CAMERA)) {
                    AlertDialog.Builder alertBuilder = new AlertDialog.Builder(context);
                    alertBuilder.setCancelable(false);
                    alertBuilder.setTitle(context.getString(R.string.permission_required));
                    alertBuilder.setMessage(context.getString(R.string.camera_permission_required));
                    alertBuilder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            ActivityCompat.requestPermissions((Activity) context, new String[]{Manifest.permission.CAMERA}, PERM_CAMERA);
                        }
                    });
                    AlertDialog alert = alertBuilder.create();
                    alert.show();
                } else {
                    ActivityCompat.requestPermissions((Activity) context, new String[]{Manifest.permission.CAMERA}, PERM_CAMERA);
                }
                return false;
            } else {
                return checkStorageWritePermission(context);
            }
        } else {
            return true;
        }
    }

    public static boolean checkStorageWritePermission(final Context context) {
        int currentAPIVersion = Build.VERSION.SDK_INT;
        if(currentAPIVersion>= Build.VERSION_CODES.M) {
            if (ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.shouldShowRequestPermissionRationale((Activity) context, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                    AlertDialog.Builder alertBuilder = new AlertDialog.Builder(context);
                    alertBuilder.setCancelable(false);
                    alertBuilder.setTitle(context.getString(R.string.permission_required));
                    alertBuilder.setMessage(context.getString(R.string.write_permission_required));
                    alertBuilder.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            ActivityCompat.requestPermissions((Activity) context, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, PERM_WRITE_EXTERNAL_STORAGE);
                        }
                    });
                    AlertDialog alert = alertBuilder.create();
                    alert.show();
                } else {
                    ActivityCompat.requestPermissions((Activity) context, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, PERM_WRITE_EXTERNAL_STORAGE);
                }
                return false;
            } else {
                return true;
            }
        } else {
            return true;
        }
    }

    public static boolean verifyPermission(int[] grantResults) {
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            return true;
        } else {
            return false;
        }
    }
}

File: picker/src/main/java/com/ds/magicpicker/PermissionUtil.kt
package com.ds.magicpicker

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat

/**
 * Created by deepaksingh on 24/05/19.
 */
val PERM_READ_EXTERNAL_STORAGE: Int = 121
val PERM_WRITE_EXTERNAL_STORAGE : Int = 122
val PERM_CAMERA : Int = 123

fun checkStorageReadPermission(context: Context): Boolean {
    val currentAPIVersion : Int = Build.VERSION.SDK_INT
    if (currentAPIVersion >= Build.VERSION_CODES.M) {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(context as Activity, Manifest.permission.READ_EXTERNAL_STORAGE)) {
                var alertBuilder = android.support.v7.app.AlertDialog.Builder(context)
                alertBuilder.setCancelable(true)
                alertBuilder.setTitle(context.getString(R.string.permission_required))
                alertBuilder.setMessage(context.getString(R.string.gallery_permission_required))
                alertBuilder.setPositiveButton(android.R.string.yes) { p0, p1 ->
                    ActivityCompat.requestPermissions(context, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), PERM_READ_EXTERNAL_STORAGE)
                }
                alertBuilder.create()
                alertBuilder.show()
            } else {
                ActivityCompat.requestPermissions(context, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), PERM_READ_EXTERNAL_STORAGE)
            }
            return false
        } else {
            return checkStorageWritePermission(context)
        }
    }
    return true
}

fun checkStorageWritePermission(context: Context): Boolean {
    val currentAPIVersion : Int = Build.VERSION.SDK_INT
    if (currentAPIVersion >= Build.VERSION_CODES.M) {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(context as Activity, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                var alertBuilder = android.support.v7.app.AlertDialog.Builder(context)
                alertBuilder.setCancelable(true)
                alertBuilder.setTitle(context.getString(R.string.permission_required))
                alertBuilder.setMessage(context.getString(R.string.write_permission_required))
                alertBuilder.setPositiveButton(android.R.string.yes) { p0, p1 ->
                    ActivityCompat.requestPermissions(context, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), PERM_WRITE_EXTERNAL_STORAGE)
                }
                alertBuilder.create()
                alertBuilder.show()
            } else {
                ActivityCompat.requestPermissions(context, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), PERM_WRITE_EXTERNAL_STORAGE)
            }
            return false
        } else {
            return true
        }
    }
    return true
}

fun checkCameraPermission(context: Context): Boolean {
    val currentAPIVersion : Int = Build.VERSION.SDK_INT
    if (currentAPIVersion >= Build.VERSION_CODES.M) {
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(context as Activity, Manifest.permission.CAMERA)) {
                var alertBuilder = android.support.v7.app.AlertDialog.Builder(context)
                alertBuilder.setCancelable(true)
                alertBuilder.setTitle(context.getString(R.string.permission_required))
                alertBuilder.setMessage(context.getString(R.string.camera_permission_required))
                alertBuilder.setPositiveButton(android.R.string.yes) { p0, p1 ->
                    ActivityCompat.requestPermissions(context, arrayOf(Manifest.permission.CAMERA), PERM_CAMERA)
                }
                alertBuilder.create()
                alertBuilder.show()
            } else {
                ActivityCompat.requestPermissions(context, arrayOf(Manifest.permission.CAMERA), PERM_CAMERA)
            }
            return false
        } else {
            return checkStorageWritePermission(context)
        }
    }
    return true
}

fun verifyPermission(grantResults: IntArray) : Boolean {
    return (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)
}



